src/cfnlint/rules/resources/iam/Sid.py

```diff
@@ -0,0 +1,95 @@
+"""
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: MIT-0
+"""
+import re
+from cfnlint.rules import CloudFormationLintRule
+from cfnlint.rules import RuleMatch
+
+
+class Sid(CloudFormationLintRule):
+    """Check if IAM Policy Sid is correct"""
+    id = 'W2512'
+    shortdesc = 'Check IAM Resource Policies Sid syntax'
+    description = 'See if the elements inside an IAM Resource policy are configured correctly.'
+    source_url = 'https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_grammar.html'
+    tags = ['properties', 'iam']
+
+    def __init__(self):
+        """Init"""
+        super(Sid, self).__init__()
+        self.resources_and_keys = {
+            'AWS::ECR::Repository': 'RepositoryPolicyText',
+            'AWS::Elasticsearch::Domain': 'AccessPolicies',
+            'AWS::KMS::Key': 'KeyPolicy',
+            'AWS::S3::BucketPolicy': 'PolicyDocument',
+            'AWS::SNS::TopicPolicy': 'PolicyDocument',
+            'AWS::SQS::QueuePolicy': 'PolicyDocument',
+        }
+        self.idp_and_keys = {
+            'AWS::IAM::Group': 'Policies',
+            'AWS::IAM::ManagedPolicy': 'PolicyDocument',
+            'AWS::IAM::Policy': 'PolicyDocument',
+            'AWS::IAM::Role': 'Policies',
+            'AWS::IAM::User': 'Policies',
+        }
+        for resource_type in self.resources_and_keys:
+            self.resource_property_types.append(resource_type)
+        for resource_type in self.idp_and_keys:
+            self.resource_property_types.append(resource_type)
+        self.sid_regex = re.compile('^[A-Za-z0-9]*$')
```

Comment:
SIDs for ManagedPolicies have a different regex than SIDs in an inline policy, IIRC. I believe spaces are allowed in an inline policy SID

Reply:
Similar discussion here: https://github.com/aws-cloudformation/cfn-python-lint/issues/708

Reply:
I started working off the cloudformation error I revieved on my stack.
Then I turned to the AWS docs to try and find more details.
The docs are pretty vague - particularly for resource policies
If I get a chance I'll set up some test templates and try them out. Seems to be where #708 got to

Reply:
I can confirm that ECR and KMS policies do not limit to that regex or require uniqueness.
I have not tested the other types but it seems that the rules are definitely different between resource types.
Given that, I think for now these rules should only apply to IAM resources.
I will update my PR to reflect that

Reply:
Still to confirm the correctness for inline policies